### PR TITLE
Feature/15885 organisation is buyer

### PR DESCRIPTION
**Pull request by MichaelRobinson-BJSS**
Added check to see if organisation is a buyer before it can be added
Added check if organisation has already been added
Fixed a number of back-links
Added related tests to OdsService and OrganisationService

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 1]
```diff
@@ -79,13 +79,16 @@ public IActionResult Find(string ods)
         }
 
         [HttpPost("find")]
-        public IActionResult Find(FindOrganisationModel model)
+        public async Task<IActionResult> Find(FindOrganisationModel model)
         {
+            (var organisation, var error) = await odsService.GetOrganisationByOdsCode(model.OdsCode);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:84]: Would probably be better either as a single deconstruction or with explicit typing, e.g.

```c#
var (organisation, error) = await odsService.GetOrganisationByOdsCode(model.OdsCode);
```

or

```c#
(OdsOrganisation organisation, string error) = await odsService.GetOrganisationByOdsCode(model.OdsCode);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 2]
```diff
@@ -95,7 +98,7 @@ public IActionResult Find(FindOrganisationModel model)
         [HttpGet("find/select")]
         public async Task<IActionResult> Select(string ods)
         {
-            var organisation = await odsService.GetOrganisationByOdsCode(ods);
+            (var organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:101]: As above, would probably be better either as a single deconstruction or with explicit typing, e.g.

```c#
var (organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```

or

```c#
(OdsOrganisation organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 3]
```diff
@@ -115,7 +118,7 @@ public IActionResult Select(SelectOrganisationModel model)
         [HttpGet("find/select/create")]
         public async Task<IActionResult> Create(string ods)
         {
-            var organisation = await odsService.GetOrganisationByOdsCode(ods);
+            (var organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:121]: As above, would probably be better either as a single deconstruction or with explicit typing, e.g.

```c#
var (organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```

or

```c#
(OdsOrganisation organisation, _) = await odsService.GetOrganisationByOdsCode(ods);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 4]
```diff
@@ -126,15 +129,30 @@ public async Task<IActionResult> Create(CreateOrganisationModel model)
             if (!ModelState.IsValid)
                 return View(model);
 
-            var organisation = await odsService.GetOrganisationByOdsCode(model.OdsOrganisation.OdsCode);
-            var orgId = await organisationsService.AddOdsOrganisation(organisation, model.CatalogueAgreementSigned);
+            (var organisation, _) = await odsService.GetOrganisationByOdsCode(model.OdsOrganisation.OdsCode);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:132]: As above, would probably be better either as a single deconstruction or with explicit typing, e.g.

```c#
var (organisation, _) = await odsService.GetOrganisationByOdsCode(model.OdsOrganisation.OdsCode);
```

or

```c#
(OdsOrganisation organisation, _) = await odsService.GetOrganisationByOdsCode(model.OdsOrganisation.OdsCode);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 5]
```diff
@@ -126,15 +129,30 @@ public async Task<IActionResult> Create(CreateOrganisationModel model)
...
+
+            (var orgId, var error) = await organisationsService.AddOdsOrganisation(organisation, model.CatalogueAgreementSigned);
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:134]: As above, would probably be better either as a single deconstruction or with explicit typing, e.g.

```c#
var (orgId, error) = await organisationsService.AddOdsOrganisation(organisation, model.CatalogueAgreementSigned);
```

or

```c#
(Guid orgId, string error) = await organisationsService.AddOdsOrganisation(organisation, model.CatalogueAgreementSigned);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/AddAnOrganisationErrorModel.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using NHSD.GPIT.BuyingCatalogue.WebApp.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Models
+{
+    public class AddAnOrganisationErrorModel : NavBaseModel
```
Comment (bjss-chris) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/AddAnOrganisationErrorModel.cs:5]: `sealed`

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+            memoryCacheMock
+                .Setup(m => m.TryGetValue($"ODS-{OdsCode}", out expectedValue))
+                .Returns(false);
+
+            memoryCacheMock.Setup(m => m.CreateEntry(It.IsAny<object>())).Returns(Mock.Of<ICacheEntry>());
+
+            var settings = new OdsSettings
+            {
+                ApiBaseUrl = "https://spineservice",
+                BuyerOrganisationRoleIds = new string[] { "RO98", "RO177", "RO213", "RO272" },
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:133]: Type name can be omitted in array initializer.

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+        public static async Task GetBuyerOrganisationByOdsCode_WithValidResponse_NotBuyerRole_ReturnsError()
+        {
+            using var httpTest = new HttpTest();
+            httpTest.RespondWith(status: 200, body: ValidResponseBody);
+
+            var memoryCacheMock = new Mock<IMemoryCache>();
+
+            object expectedValue = null;
+            memoryCacheMock
+                .Setup(x => x.TryGetValue($"ODS-{OdsCode}", out expectedValue))
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:157]: Lazy lambda parameter name.

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+                .Returns(false);
+
+            memoryCacheMock.Setup(m => m.CreateEntry(It.IsAny<object>())).Returns(Mock.Of<ICacheEntry>());
+
+            var settings = new OdsSettings
+            {
+                ApiBaseUrl = "https://spineservice",
+                BuyerOrganisationRoleIds = new string[] { "X123" },
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:165]: As above, type name can be omitted.

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+            result.Should().BeEquivalentTo(serviceRecipients);
+            memoryCacheMock.Verify(v => v.CreateEntry(It.IsAny<object>()), Times.Never);
+        }
+
+        [Fact]
+        public static async Task GetServiceRecipientsByParentOdsCode_SinglePage_ReturnsOrganisation()
+        {
+            var odsCode = "123";
+
+            var childOrg = new ServiceRecipient { Name = "Organisation 1", PrimaryRoleId = "RO177", OrgId = "ABC" };
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:245]: Can we create `odsCode` and `childOrg` using AutoFixture?

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+            memoryCacheMock.Verify(v => v.CreateEntry(It.IsAny<object>()), Times.Once);
+        }
+
+        [Fact]
+        public static async Task GetServiceRecipientsByParentOdsCode_MultiplePages_ReturnsOrganisation()
+        {
+            var odsCode = "123";
+
+            var childOne = new ServiceRecipient { Name = "Organisation 1", PrimaryRoleId = "RO177", OrgId = "ABC" };
+            var childTwo = new ServiceRecipient { Name = "Organisation 2", PrimaryRoleId = "RO177", OrgId = "ABD" };
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:281]: As above, can we create `odsCode`, `childOne` and `childTwo` using AutoFixture?

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+            memoryCacheMock.Verify(v => v.CreateEntry(It.IsAny<object>()), Times.Once);
+        }
+
+        [Fact]
+        public static async Task GetServiceRecipientsByParentOdsCode_SinglePageDifferentRoleIds_ReturnsOnlyMatching()
+        {
+            var odsCode = "123";
+
+            var childOne = new ServiceRecipient { Name = "Organisation 1", PrimaryRoleId = "RO177", OrgId = "ABC" };
+            var childTwo = new ServiceRecipient { Name = "Organisation 2", PrimaryRoleId = "RO178", OrgId = "ABD" };
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:321]: As above, can we create `odsCode`, `childOne` and `childTwo` using AutoFixture?

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+            var result = await service.GetServiceRecipientsByParentOdsCode(odsCode);
+
+            result.Should().BeEquivalentTo(childOne);
+            memoryCacheMock.Verify(v => v.CreateEntry(It.IsAny<object>()), Times.Once);
+        }
+
+        [Fact]
+        public static async Task GetServiceRecipientsByParentOdsCode_NoOrganisations_ReturnsEmptyList()
+        {
+            var odsCode = "123";
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:354]: As above, can we create `odsCode` using AutoFixture?

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OrganisationsServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        [CommonAutoData]
+        public static async Task AddOdsOrganisation_OrganisationAlreadyExists_ReturnsError(
+            OdsOrganisation odsOrganisation,
+            bool agreementSigned,
+            Organisation organisation,
+            [Frozen] Mock<IDbRepository<Organisation, BuyingCatalogueDbContext>> repositoryMock,
+            OrganisationsService service)
+        {
+            repositoryMock.Setup(r => r.GetAllAsync(o => o.OdsCode == odsOrganisation.OdsCode))
+                .ReturnsAsync(new Organisation[] { organisation });
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OrganisationsServiceTests.cs:40]: Type name can be omitted from array initializer.

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs (hunk 1 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:84]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs (hunk 2 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:101]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs (hunk 4 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:132]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OrganisationsServiceTests.cs (hunk 14 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OrganisationsServiceTests.cs:40]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs (hunk 3 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:121]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs (hunk 5 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:134]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 8 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:157]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 9 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:165]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 10 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:245]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 12 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:321]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 7 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:133]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 11 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:281]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs (hunk 13 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Organisations/OdsServiceTests.cs:354]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/AddAnOrganisationErrorModel.cs (hunk 6 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Models/AddAnOrganisationErrorModel.cs:5]: Done

**Pull request merged**